src/yaml/from_yaml.cr

```diff
@@ -22,39 +22,59 @@ end
 
 def Nil.new(pull : YAML::PullParser)
   location = pull.location
-  value = pull.read_scalar
-  if value.empty?
+  value = pull.read_plain_scalar
+  if YAML.reserved_scalar(value) == "null"
     nil
   else
     raise YAML::ParseException.new("Expected nil, not #{value}", *location)
   end
 end
 
 def Bool.new(pull : YAML::PullParser)
-  pull.read_scalar == "true"
+  location = pull.location
+  value = pull.read_plain_scalar
+  if YAML.reserved_scalar(value) == "true"
+    true
+  elsif YAML.reserved_scalar(value) == "false"
+    false
+  else
+    raise YAML::ParseException.new("Expected boolean, not #{value}", *location)
+  end
 end
 
 {% for type in %w(Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64) %}
   def {{type.id}}.new(pull : YAML::PullParser)
     location = pull.location
     begin
-      {{type.id}}.new(pull.read_scalar)
+      new(pull.read_plain_scalar)
     rescue ex
       raise YAML::ParseException.new(ex.message.not_nil!, *location)
     end
   end
 {% end %}
 
-def String.new(pull : YAML::PullParser)
-  pull.read_scalar
-end
-
-def Float32.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f32
-end
+{% for type in %w(Float32 Float64) %}
+  def {{type.id}}.new(pull : YAML::PullParser)
+    location = pull.location
+    begin
+      value = pull.read_plain_scalar
+      return NAN if YAML.reserved_scalar(value) == "NaN"
+      return INFINITY if YAML.reserved_scalar(value) == "infinity"
+      return -INFINITY if YAML.reserved_scalar(value) == "-infinity"
+      new(value)
+    rescue ex
+      raise YAML::ParseException.new(ex.message.not_nil!, *location)
+    end
+  end
+{% end %}
 
-def Float64.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f64
+def String.new(pull : YAML::PullParser) : String
+  location = pull.location
+  style = pull.data.scalar.style
+  value = pull.read_scalar
+  return value unless style == LibYAML::ScalarStyle::PLAIN
```

Comment:
@ysbaddaden incorrect, 1.0 is reserved and is not a string, it can only be a float. I can give you numerous examples from other languages that have properly implemented yaml. Even ruby treats this accordingly.

```
$ ruby -r yaml -e 'puts YAML.load("1.0").class'
Float
$ ruby -r yaml -e 'puts YAML.load(%("1.0")).class'
String
```